File: backend/Services/Parsers/ConnectionsParser.cs
using System.Text.RegularExpressions;
using backend.Models;

namespace backend.Services.Parsers
{
    public class ConnectionsParser : BasicScoreResultParser
    {
        private ILogger<ConnectionsParser> _logger;

        public ConnectionsParser(ILogger<ConnectionsParser> logger) : base(logger)
        {
            _logger = logger;
        }

        public override bool CountWinner => true;
        public override bool Default => false;
        protected override string? ExtraContent => null;
        public override string GameName => "Connections";
        public override string? HelpText => null;

        protected override Regex Parser => new Regex($"{GameName}\nPuzzle #[\\d]+\n(?<{ScoreGroup}>.*)");
        public override string Url => "https://www.nytimes.com/games/connections";

        public override string? GetScoreValue(DailyResult dailyResult)
        {
            return dailyResult.Scores == null || dailyResult.Scores.Count != 4 ? null : dailyResult.Scores.Average().ToString();
        }

        protected override DailyResult SetScore(DailyResult dailyResult, Match parserResults)
        {
            var scoreRows = parserResults.Groups[ScoreGroup]
                .Value
                .Split('\n');

            var totalGuesses = scoreRows.Length;

            var correctRows = scoreRows
                .Where(row => row.Distinct().Count() == 1)
                .Count();

            // the score is the total number of guesses plus the number of categories that were not successfully guessed
            dailyResult.Score = totalGuesses + (4 - correctRows);

            return dailyResult;
        }
    }
}

File: lambdas/src/parsers/connections-parser.ts
import { DailyResult } from '../models/daily-result';
import { BasicScoreResultParser } from './base/basic-score-parser';
import { getGroup, hasGroup } from './base/result-parser';

export class ConnectionsParser extends BasicScoreResultParser {
  readonly countWinner = true;
  readonly gameName = 'Connections';
  readonly helpText = null;
  protected readonly extraContent = null;
  readonly url = 'https://www.nytimes.com/games/connections';
  
  protected readonly parser = new RegExp(
    `^${this.gameName}\\s*Puzzle #\\d+(?<${BasicScoreResultParser.SCORE_GROUP}>.*)$`,
    's'
  );
  
  protected getCleanResult(result: string, match: RegExpMatchArray): string {
    return result;
  }
  
  protected setScore(dailyResult: DailyResult, match: RegExpMatchArray): DailyResult {
    if (!hasGroup(match, BasicScoreResultParser.SCORE_GROUP)) {
      return dailyResult;
    }
    
    const scoreContent = getGroup(match, BasicScoreResultParser.SCORE_GROUP) || '';
    const scoreRows = scoreContent
      .split('\n')
      .filter(row => row.trim().length > 0)
      .map(row => row.trim());
    
    const totalGuesses = scoreRows.length;
    
    // Count correct rows (all same emoji)
    const correctRows = scoreRows.filter(row => {
      const parts: string[] = [];
      for (let i = 0; i < row.length - 1; i += 2) {
        parts.push(row.substring(i, i + 2));
      }
      const uniqueParts = [...new Set(parts)];
      return uniqueParts.length === 1;
    }).length;
    
    // Score is total guesses plus missed categories
    dailyResult.score = totalGuesses + (4 - correctRows);
    
    return dailyResult;
  }
}
